calico-vpp-agent/routing/routing_server.go

```diff
@@ -166,33 +190,36 @@ func (s *Server) serveOne() error {
 	s.RestoreLocalAddresses()
 
 	/* We should start watching from our GetNodeIPs change call */
-	s.t.Go(func() error { return s.nodeWatcher.WatchNodes(node.ResourceVersion) })
+	s.t.Go(func() error { return s.nodeWatcher.WatchNodes(node.ResourceVersion, s.t.Dying()) })
 	// sync IPAM and call ipamUpdateHandler
-	s.t.Go(func() error { return s.ipam.SyncIPAM() })
+	s.t.Go(func() error { return s.ipam.SyncIPAM(s.t.Dying()) })
 	// watch prefix assigned and announce to other BGP peers
-	s.t.Go(func() error { return s.prefixWatcher.WatchPrefix() })
+	s.t.Go(func() error { return s.prefixWatcher.WatchPrefix(s.t.Dying()) })
 	// watch BGP peers
-	s.t.Go(func() error { return s.peerWatcher.WatchBGPPeers() })
+	s.t.Go(func() error { return s.peerWatcher.WatchBGPPeers(s.t.Dying()) })
 	// watch Felix configuration
-	s.t.Go(func() error { return s.felixConfWatcher.WatchFelixConfiguration() })
+	s.t.Go(func() error { return s.felixConfWatcher.WatchFelixConfiguration(s.t.Dying()) })
 
 	// TODO need to watch BGP configurations and restart in case of changes
 	// Need to get initial BGP config here, pass it to the watchers that need it,
 	// and pass its revision to the BGP config and nodes watchers
 
 	// watch routes from other BGP peers and update FIB
-	s.t.Go(func() error { return s.bgpWatcher.WatchBGPPath() })
-	s.t.Go(func() error { return s.connectivityServer.ServeConnectivity() })
+	s.t.Go(func() error { return s.bgpWatcher.WatchBGPPath(s.t.Dying()) })
+	s.t.Go(func() error { return s.connectivityServer.ServeConnectivity(s.t.Dying()) })
+	s.t.Go(func() error { return s.WatchBGPConf(s.t.Dying()) })
 
 	// watch routes added by kernel and announce to other BGP peers
 	// FIXME : s.t.Go(s.kernelWatcher.WatchKernelRoute())
 
 	s.ipam.WaitReady()
 	ServerRunning <- 1
-	s.log.Infof("Routing server is running ")
-	<-s.t.Dying()
-	s.log.Warnf("routing tomb returned %v", s.t.Err())
-
+	select {
+	case <-s.mainTomb.Dying():
```

Comment:
When the main Tomb signals Dying, we want to cleanly stop all the watchers by stopping `s.t`, not just return.
If `s.t` exits first with an error, it could be nice to eventually return this error.